Rate limiting skeleton

aaron-prindle opened the pull request:
Filling out test skeleton

lavalamp committed:
Merge pull request #76902 from aaron-prindle/feature-rate-limiting

[FEATURE BRANCH] Rebase 'feature-rate-limiting' branch w/ master

lavalamp committed:
Merge pull request #77500 from yue9944882/chore/20190506-rebase-feature-rate-limiting

[FEATURE BRANCH] Rebase 'feature-rate-limiting' branch w/ master

yue9944882 committed:
update feature approver

k8s-ci-robot committed:
Merge pull request #77652 from yue9944882/chore/feature-top-approver

[FEATURE BRANCH] Add feature-only top approvers

MikeSpreitzer committed:
Started drafting skeleton of the request management filter

yue9944882 committed:
add developers to api-approver

aaron-prindle committed:
implementation working with simple test setup.  dynamic configuration modified to be static for testing.

lavalamp committed:
Merge pull request #78014 from yue9944882/20190517-rebase-feature-rate-limiting

[FEATURE BRANCH] Rebase feature rate limiting w/ master

k8s-ci-robot committed:
Merge pull request #79420 from yue9944882/chore/feature-api-approver

[FEATURE BRANCH] Add developers to api-approver to unblock people

yue9944882 committed:
versioned flow-control api model

yue9944882 committed:
[DO NOT REVIEW]: trival codes

yue9944882 committed:
add pkg owners: unblock feature-branch rebase

lavalamp committed:
Merge pull request #80170 from yue9944882/chore/add-more-pkg-owner

[FEATURE BRANCH] Add more no-parent pkg owner to unblock rebasing

lavalamp committed:
Merge pull request #80159 from yue9944882/chore/20190715-rebase-feature-rate-limiting

[FEATURE BRANCH] Rebase feature rate limiting w/ master

k8s-ci-robot committed:
Merge pull request #77048 from yue9944882/feat/rate-limiting-api-model

[FEATURE BRANCH]: flowcontrol.apiserver.k8s.io initial API models

MikeSpreitzer committed:
update bazel by hack

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/server/filters/reqmgmt.go:
```diff
@@ -47,11 +52,10 @@ import (
 // FairQueuingFactory knows how to make FairQueueingSystem objects.
 // This filter makes a FairQueuingSystem for each priority level.
 type FairQueuingFactory interface {
-	NewFairQueuingSystem(concurrencyLimit, numQueues, queueLengthLimit int, requestWaitLimit time.Duration, clk clock.Clock) FairQueuingSystem
```
The `concurrencyLimit` parameter should be there!

aaron-prindle replied:
I've added concurrencyLimit back into the constructor and wired it up correctly for the static case (no dynamic configs).

MikeSpreitzer committed:
Expanded comment on FairQueuingSystem::DoOnEmpty

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/server/filters/fq/fq.go:
```diff
@@ -0,0 +1,178 @@
+	robinidx     int
+}
+
+func (q *FQScheduler) chooseQueue(packet FQPacket) FQQueue {
```
The variable name "q" suggests, to me, "queue" more than "scheduler". What about using a name like "fqs" or "sch" here instead of "q"?

aaron-prindle replied:
updated

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/server/filters/fq/fq.go:
```diff
@@ -0,0 +1,178 @@
+
+	// When a request finishes being served, and the actual service time was S,
+	// the queue’s virtual start time is decremented by G - S.
+	virstart := q.Queues[p.GetQueueIdx()].GetVirStart()
```
What would you think of computing `q.Queues[p.GetQueueIdx()]` just once and storing the result in a local variable?

aaron-prindle replied:
updated

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/server/filters/reqmgmt.go:
```diff
@@ -76,9 +80,231 @@ type FairQueuingSystem interface {
+		packets := []*fq.Packet{}
+		fqpackets := make([]fq.FQPacket, len(packets), len(packets))
+		for i := range packets {
+			fqpackets[i] = fqpackets[i]
```
This statement is a no-op.

aaron-prindle replied:
Sorry for not catching this, it was not intentional.  I've removed this

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/server/filters/reqmgmt.go:
```diff
@@ -76,9 +80,231 @@ type FairQueuingSystem interface {
+		queues = append(queues, &fq.Queue{})
+		packets := []*fq.Packet{}
+		fqpackets := make([]fq.FQPacket, len(packets), len(packets))
+		for i := range packets {
```
This loop is guaranteed to iterate zero times.  Why is it here?

aaron-prindle replied:
removed

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/server/filters/reqmgmt.go:
```diff
@@ -76,9 +80,231 @@ type FairQueuingSystem interface {
+		queues[i].Packets = fqpackets
+	}
+
+	fqqueues := make([]fq.FQQueue, len(queues), len(queues))
```
`len(queues)` equals `numQueues`; this statement could come much earlier.

aaron-prindle replied:
updated

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/server/filters/reqmgmt.go:
```diff
@@ -76,9 +80,231 @@ type FairQueuingSystem interface {
+	}
+
+	fqqueues := make([]fq.FQQueue, len(queues), len(queues))
+	for i := range queues {
```
This loop could be merged into the previous one.

aaron-prindle replied:
updated

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/server/filters/reqmgmt.go:
```diff
@@ -76,9 +80,231 @@ type FairQueuingSystem interface {
+			limit := channelPkt.EnqueueTime.Add(s.requestWaitLimit)
+			if now.After(limit) {
+
+				channelPkt.DequeueChannel <- false
```
At this point the packet needs to be removed from the queue.  And the channel should probably be closed after the send.

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/server/filters/reqmgmt.go:
```diff
@@ -88,6 +314,17 @@ type RMState struct {
+		denom += int(pl.config.AssuredConcurrencyShares)
+		// denom += ACS(prioritylvl, queues)
+	}
+	ps.concurrencyLimit = int(math.Ceil(float64(r.serverConcurrencyLimit * int(ps.config.AssuredConcurrencyShares) / denom)))
```
This statement does an integer division; we want a floating point division.  I suggest starting with `var denom float64` and adding `float64(pl.config.AssuredConcurrencyShares)` for each `pl`.

aaron-prindle replied:
updated

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/server/filters/reqmgmt.go:
```diff
@@ -76,9 +80,231 @@ type FairQueuingSystem interface {
 // RMState is the variable state that this filter is working with at a
 // given point in time.
 type RMState struct {
+	// TODO(aaron-prindle) remove?
```
No, this belongs here.

aaron-prindle replied:
updated, comment removed

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/server/filters/reqmgmt.go:
```diff
@@ -149,7 +386,7 @@ type requestManagement struct {
 
 // TypedConfigObjectReference is a reference to a relevant config API object.
-// No namespace is needed because none of these objects is namespaced.
+// No namespace is needed because none of these objects is namespaces.
```
"namespaced" was correct.

aaron-prindle replied:
Sorry, I'm not sure how these comments were changed I agree with "namespaced".  I will change them back.

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/server/filters/reqmgmt.go:
```diff
@@ -159,84 +396,192 @@ func (tr *TypedConfigObjectReference) String() string {
-			handler.ServeHTTP(w, r)
-			return
-		}
+		// if longRunningRequestCheck != nil && longRunningRequestCheck(r, requestInfo) {
```
Why comment this out?  It is easy enough to pass `nil` in a test.

aaron-prindle replied:
updated

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/server/filters/reqmgmt.go:
```diff
@@ -159,84 +396,192 @@ func (tr *TypedConfigObjectReference) String() string {
+
+		// TODO(aaron-prindle) verify this is needed?
+		rmState.updateACV(ps) // TODO(aaron-prindle) rename updateACV?
+		ps.fqs.SetConcurrencyLimit(ps.concurrencyLimit)
```
This does not need to be done at every request.  It needs to be done when the configuration changes, and during initial setup.  I do not understand why you removed the `concurrencyLimit` parameter from the FairQueueingSystem constructor.

aaron-prindle replied:
I originally did not understand why it would be set as I thought it was dynamic from rmState now I believe I understand the intention of concurrencyLimit.  This is updated, `concurrencyLimit` is again in the FairQueueingSystem constructor

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/server/filters/reqmgmt.go:
```diff
@@ -159,84 +396,192 @@ func (tr *TypedConfigObjectReference) String() string {
 		} else {
 			klog.V(5).Infof("Rejecting %v\n", r)
-			tooManyRequests(r, w)
+			// tooManyRequests(r, w) // TODO(aaron-prindle)
```
Why comment this out?  It invokes an existing function.

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/server/filters/reqmgmt.go:
```diff
@@ -159,84 +396,192 @@ func (tr *TypedConfigObjectReference) String() string {
 	// TODO: implement
-	return nil
+	// TODO(aaron-prindle) verify implementation
+	return hash(fmt.Sprintf("%s,%s", fsName, fDistinguisher))
```
Since `hash.Hash` has a `Write` method that can be invoked multiple times, we do not need to pay to construct a string here.

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/server/filters/reqmgmt/inflight/fq.go:
```diff
@@ -0,0 +1,140 @@
+
+	select {
+	case reqMgmtPkt.DequeueChannel <- true:
+	default:
```
For every packet, the total number of sends over its lifetime should be exactly 1 --- in any scenario.  We do not need three layers of abstraction, each with their own mutex.  We only need only layer, the FairQueueingSystem implementation, and with only one mutex, and no uncertainty about whether the packet's channel has yet be sent to.  There also needs to be exactly 1 close of that channel, and that will not be difficult if we have just one layer.

aaron-prindle replied:
I agree, the current levels of abstraction are from the history of the implementation and helped iterate faster until this point.  I will condense them into a single layer.

aaron-prindle replied:
I am waiting to make this change in a separate PR, currently there is a test suite, fq_test.go which has some tests around fairness for the underlying FQ implementation.  The abstraction levels help verify that those tests are unchanged as I build on top of it.  Once there is consensus on the approach here I can refactor into a single layer, it should be simple.

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/server/filters/reqmgmt_test.go:
```diff
@@ -0,0 +1,439 @@
+			},
+		)
+	}
+	// can't init nested struct in initializer
```
Huh?
```
rmtypesv1a1.PriorityLevelConfiguration{
    ObjectMeta: metav1.ObjectMeta{Name: fmt.Sprintf("plc-%d", i)},
    Spec: rmtypesv1a1.PriorityLevelConfigurationSpec{ ... }
}
```

aaron-prindle replied:
updated

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/server/filters/reqmgmt_test.go:
```diff
@@ -0,0 +1,439 @@
+	counts := []int64{}
+	for range flowSchemas {
+		counts = append(counts, int64(0))
+	}
```
Simpler: `counts := make([]int64, len(flowSchemas) )`

aaron-prindle replied:
updated

MikeSpreitzer reviewed (commented):
This is painful to review because there are so many files changed.  But they are not really part of the change here; a bunch of old merged stuff has been dragged in because this is not based on the latest merged code.  Also, rebasing on the latest https://github.com/kubernetes/kubernetes/pull/78966 would pick up relevant improvements there and avoid the chaos of forked development.

I see three layers where we only need one: an implementation of FairQueueingSystem.  We only need a mutex per FairQueuingSystem, plus one more at the `requestManagement` to update its `curState` in response to configuration changes (ooh, maybe not even a mutex there --- just let the config queue have a single worker).

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/server/filters/reqmgmt_test.go:
```diff
@@ -0,0 +1,439 @@
+	var count int64
+	delegate := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		atomic.AddInt64(&count, int64(1))
+		time.Sleep(1 * time.Second)
```
We should not use real time in tests.  We should arrange our tests so that a fake clock can be used, and advanced as soon as the computation for each time is finished.
See https://github.com/MikeSpreitzer/kubernetes/blob/add-event-clock/staging/src/k8s.io/apimachinery/pkg/util/clock/event_clock.go for a clock that makes that easy.

aaron-prindle replied:
I have not completed this change but definitely agree and am in the process of making testing use a fake clock.

aaron-prindle committed:
addressing comments

MikeSpreitzer committed:
A little more thought about concurrency in the reqmgmt filter

(1) removed the idea of multiple config queue workers; it will be
simpler and adequate if there is only 1.

(2) attempted to clarify the comment on FairQueuingSystem::DoOnEmpty
to (a) make the division of responsibilities clearer, (b) make a
precise statement of the semantics, and (c) be less prescriptive of
how the filter uses this method.
